### PR TITLE
[Proxying] Add `emscripten_proxy_callback_with_ctx`

### PR DESCRIPTION
**Pull request by tlively**
This new proxying function is like `emscripten_proxy_callback`, but supports
proxying of asynchronous work just like `emscripten_proxy_sync_with_ctx`. It
uses the existing `em_proxying_ctx` type and `emscripten_proxy_finish` function
to mark the work done, and that type and function are internally augmented to
handle both synchronous and callback-based proxying. `emscripten_proxy_callback`
is reimplemented to share most of its implementation with
`emscripten_proxy_callback_with_ctx`, but it does break the abstraction slightly
to avoid having to make an extra allocation.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #18741** <a href="https://app.graphite.dev/github/pr/emscripten-core/emscripten/18741" target="_blank">Graphite</a> 
    * **PR #18776** <a href="https://app.graphite.dev/github/pr/emscripten-core/emscripten/18776" target="_blank">Graphite</a> 
      * **PR #18810** <a href="https://app.graphite.dev/github/pr/emscripten-core/emscripten/18810" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/emscripten-core/emscripten/18810?utm_source=stack-comment).

**Comment by tlively** [comment]
@sbc100 this is good to go besides review.

File: system/include/emscripten/proxying.h [hunk 1]
```diff
@@ -88,6 +88,20 @@ int emscripten_proxy_callback(em_proxying_queue* q,
...
                               void* arg);
 
+// Enqueue `func` on the given queue and thread. Once (and if) it finishes the
+// task by calling `emscripten_proxy_finish` on the given `em_proxying_ctx`, it
+// will asynchronously proxy `callback` back to the current thread on the same
+// queue, or if the target thread dies before the work can be completed,
+// `cancel` will be proxied back instead. All three function will receive the
+// same argument, `arg`. Returns 1 if `func` was successfully enqueued and the
+// target thread notified or 0 otherwise.
+int emscripten_proxy_callback_with_ctx(em_proxying_queue* q,
```
Comment (sbc100) [inline system/include/emscripten/proxying.h:98]: Should we just called this  `emscripten_proxy_async_with_ctx`?

I think maybe the `_with_callback` distinction is important, and also obvious the from the args.    Ideally I think we could remove the `_with_callback` variants that just have the normal versions take optional callbacks in all cases.

Another reason is that all the current function start with either `emscripten_proxy_sync` or `emscripten_proxy_async`

**Review by sbc100 — COMMENTED** [review]
lgtm % naming .. which we can bikeshed later.

File: system/include/emscripten/proxying.h (hunk 1 above)
Comment (tlively) [reply system/include/emscripten/proxying.h:98]: After this full sequence of PRs, we have :

```
emscripten_proxy_async
emscripten_proxy_sync{_with_ctx}
emscripten_proxy_callback{_with_ctx}
emscripten_proxy_promise{_with_ctx}
```

If we really wanted to minimize the API surface, my preference would be to remove `emscripten_proxy_async` and `emscripten_proxy_callback*`, leaving only `sync` and `promise` proxying, with and without `ctx`s. The reason I haven't tried to do that minimization yet is that it seems low cost to expose the extra functions if we're going to have nearly identical functions under the hood either way. Users get the most options at negligible maintenance cost to us.

Comment (sbc100) [reply system/include/emscripten/proxying.h:98]: Sounds reasonable.

As a general rule, I would rather keep the low level C API as small as possible.  For this kind of low level API (which I hope most users will never need to use), I think its fine if we don't include helpers/wrappers/shortcuts.

If we want to provide them in C++ API, or in a higher level set of C APIS, that would be fine, but at least for me it helper conceptually to keep this surface small... but maybe thats just me.

None of this has to block this change..  I don't think refactoring this layer later on is very risky.

**Review by sbc100 — APPROVED** [review]

**Pull request merged**